### PR TITLE
Add shader-float16 extension

### PR DESCRIPTION
**Pull request by xhcao**
@haoxli

***
<a href="https://pr-preview.s3.amazonaws.com/xhcao/gpuweb/pull/1275.html" title="Last updated on Dec 4, 2020, 7:33 AM UTC (9bfa65a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1275/b5b20b6...xhcao:9bfa65a.html" title="Last updated on Dec 4, 2020, 7:33 AM UTC (9bfa65a)">Diff</a>

### CODE REVIEW

**Comment by xhcao** [comment]
@Kangz @kainino0x  PTAL

**Comment by xhcao** [comment]
Sorry, I had known that it was not so easy to introduce an extension as this pull request. Please ignore this pull request if it is not appropriate.
The goal was that I encountered an issue when using shader-float16 extension in tfjs. The "webgpu/types" defines GPUExtensionName as shown below,
  export type GPUExtensionName =
    | "texture-compression-bc"
    | "timestamp-query"
    | "pipeline-statistics-query"
    | "depth-clamping";
There is no "shader-float16", so if I query whether device supports fp16 features as shown below, compiling error happens.
  const adapter = await navigator.gpu.requestAdapter(gpuDescriptor);
  let driverSupportFp16 = false;
  for(let i=0; i<adapter.extensions.length; i++){
    if (adapter.extensions[i] === 'shader-float16')
    {
      driverSupportFp16 = true;
      break;
    }
  }

**Comment by Kangz** [comment]
It's good to put this extension on the radar of the group, but I don't know if we'll want to take the extension name in without any specification. We'll discuss it on monday's meeting. In the meantime you can use Typescript's features to cast and ignore the type error.

File: spec/index.bs [hunk 1]
```diff
@@ -7156,6 +7157,13 @@ The following enums are supported if and only if the {{GPUFeatureName/"pipeline-
 </dl>
 
 
+## <dfn dfn-type=enum-value dfn-for=GPUFeatureName>shader-float16</dfn> ## {#shader-float16}
+
+Allows 16bit float arithmetic feature in shader and 16bit storage features for GPUBuffer.
```
Comment (kvark) [inline spec/index.bs:7162]: what are the "16bit storage features of GPUBuffer"?

**Review by kvark — COMMENTED** [review]
I believe this is much needed. Hopefully Tint can prototype this and see if anything weird pops up.

File: spec/index.bs (hunk 1 above)
Comment (Kangz) [reply spec/index.bs:7162]: I think it means accessing buffer memory through 16bit types in WGSL. (but it needs to be reworded to say that)

Comment (dneto0) [reply spec/index.bs:7162]: In principle this is a fine idea.

As noted, there's a lot of detail to be determined.  Support in the field is very uneven, and Vulkan slices the functionality into lots of pieces, e.g.
- some devices can compute in fp16, but not load or store individual fp16 values.
- some devices can load or store individual fp16 values but not directly do fp16 arithmetic.
- ability to load/store fp16 values differs by storage class.  storage buffer, uniform buffer, input/output, workgroup storage.

- Also, you'd need to specify all the places f16 can be used, and behaviour of arithmetic, and error bounds on builtins (if they apply)

**Review by jdashg — CHANGES_REQUESTED** [review]
Can all the WGSL changes be included in this PR?

**Comment by grorg** [comment]
Discussed at the [2020-12-07 meeting](https://docs.google.com/document/d/11VNsaOHObp5zmQ6E4DIZrQmYZdSJnwQ_JHhqEkYeFUg/edit).

**Comment by xhcao** [comment]
Hi, all, thank you for your comments. But we currently have no enough resource to implement fp16 extension in Tint project.
I had adopted Corentin's comment to work around my issue with Typescript's features.

**Comment by kainino0x** [comment]
Closing; superseded by #2696

**Pull request closed without merging**